### PR TITLE
Adding global email model

### PR DESCRIPTION
**Pull request by ehfeng**
Because emails are not distinct per `User`, we're using the `Email` to store email-specific settings (email subscription state). `UserEmail` will represent verification that a given user "owns" an email address.

Alternatively, `UserEmail.email` could just be a key for `Email.id`, where the id value is the email string. The value for this seems minimal (`UserEmail.email` is never updated in place, so the chance of `UserEmail.email != UserEmail.global_email.email` is low), so I've done the more naive approach.

### CODE REVIEW

File: src/sentry/models/useremail.py
Comment (dcramer) [inline src/sentry/models/useremail.py]: Probably don't need this since you can just index the email string and query on it

File: src/sentry/south_migrations/0346_backfill_email.py
Comment (mattrobenolt) [inline src/sentry/south_migrations/0346_backfill_email.py]: Don't do this `.save()` Do the `UserEmail.objects.filter(id=useremail.id).update(global_email=email)` pattern instead to avoid race condition.

**Review by mattrobenolt — CHANGES_REQUESTED** [review]
We definitely need to ship the data migration after the schema migration, otherwise you'll end up with some rows that are still NULL.

It should also be considered if we care about case sensitivity here or not. I feel like we should always lowercase the value to prevent unintended duplicates due to case differences. Matt@example.com vs matt@example.com

And, what is going to keep this up to date for new rows? I don't see any code that's adding new Email rows except for the migration.

**Comment by ehfeng** [comment]
@mattrobenolt good point. 

I'm looking at `receivers/useremail.py` and I'm assuming this is a good case for having a receiver on `post_save`. Especially as changes to `User` and `UserEmail` are generally user-triggered, so adding latency isn't bad?

**Comment by ehfeng** [comment]
@mattrobenolt I could modify it such that the host is case insensitive, but the username part of email [can be case-sensitive](https://stackoverflow.com/questions/9807909/are-email-addresses-case-sensitive). Obviously, Gmail, Yahoo, and Outlook make the username case-insensitive, so we could be the same, but email standard itself mandates case sensitivity.

I'm not sure if this would be a security issue—I'll defer to whatever you decide.

**Comment by mattrobenolt** [comment]
Yeah, internally @tkaemming brought this up. But our own behavior for normal `User.email` is to treat it as case-insensitive and we use this for logins. My concern is that we don't 100% handle this everywhere correctly since it requires explicit action to handle this behavior.

If we have control over this, we can just apply a functional index that would handle uniqueness as case insensitive to enforce it correctly.

**Comment by dcramer** [comment]
Can someone pull data on if we have case duplicated emails? if not lets move towards case insensitive.

**Comment by dcramer** [comment]
Let's do a few things:

1. Email.email is unique
2. Email.email is stored in LOWERCASE at all times (i dont know if this is easy to do via a custom field?)
3. UserEmail has a unique constraint on (user_id, email)
4. UserEmail.email is stored same as Email.email (always lowercase)
5. Migrate all existing emails in UserEmail to be lowercase

**Comment by mattrobenolt** [comment]
> Email.email is stored in LOWERCASE at all times (i dont know if this is easy to do via a custom field?)

I was going to do this with a functional index to enforce it on the database. Something like, `CREATE UNIQUE INDEX ... (lower(email))` since we run custom SQL anyways for indexes.

**Comment by dcramer** [comment]
@mattrobenolt is there a good reason to not store things lowercase? its a lot easier to be compatible if we do that

**Comment by mattrobenolt** [comment]
But yeah, we can do a field as well, I think having the index enforce it makes it more explicit on the db side.

I think my concerns are how we have to deal with User.email today. We have to always make sure we do a case insensitive filter, etc. Which may be error prone.

I don't think a field would help with this case, would it? For example, doing `Email.objects.filter(email='FOO@EXAMPLE.COM')` vs `Email.objects.filter(email__iexact='FOO@EXAMPLE.COM')` Not sure if a custom field could coerce this behavior automatically? I know the db index would.

**Comment by dcramer** [comment]
An index wouldn't coerce a filter clause, but a field could. We should see if we can do that, though someone from @getsentry/platform might need to help here.

**Comment by ehfeng** [comment]
We can't assume all email servers are case insensitive. If example.com's email is, then someone with the address `UserName@example.com` would literally never be able to get our emails.

I prefer Matt's suggestion of enforcing in the index, but storing the case-sensitive value.

**Comment by dcramer** [comment]
I'd prefer a case insensitive text field, but given we dont have that, I'm ok with saiyng "we will not support case sensitive email servers".

**Comment by ehfeng** [comment]
Adding [this](https://softwareengineering.stackexchange.com/questions/128297/programming-around-email-address-case-sensitivity) as reference:

> While the above definition for Local-part is relatively permissive, for maximum interoperability, a host that expects to receive mail SHOULD avoid defining mailboxes where the Local-part requires (or uses) the Quoted-string form or where the Local-part is case-sensitive.

**Comment by dcramer** [comment]
Per talk internally, let's just make it a 'citext' field for postgres, and assume its case insensitive.

In MySQL we might be able to use a case insensitive collation, but im not sure if Django lets us provide that. Ensuring correctness for SQLite would be nice, but i'm not sure we have an option.

File: src/sentry/models/email.py
Comment (mattrobenolt) [inline src/sentry/models/email.py]: Don't we need `unique=True` here?

**Review by mattrobenolt — CHANGES_REQUESTED** [review]
We still need to move the backfill to another deploy. The problem is there will be a race condition. We'll deploy the new model, run the backfill, then between running the backfill and new code wtih the signal being live, we'll miss data.

File: src/sentry/models/email.py
Comment (ehfeng) [reply src/sentry/models/email.py]: That's true. Added.

**Comment by ehfeng** [comment]
Removed backfill

**Comment by ehfeng** [comment]
@mattrobenolt We need to create the citext extension within the test database. Can you help with this?

**Comment by getsentry-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Changes require @getsentry/security sign-off</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR includes migrations</td>
    </tr>
  </tbody>
</table>

### Security concerns found

- src/sentry/models/email.py
- src/sentry/receivers/email.py


## Migration Checklist

- [ ] new columns need to be nullable (unless table is new)
- [ ] migration with any new index needs to be done concurrently
- [ ] data migrations should not be done inside a transaction
- [ ] before merging, check to make sure there aren't conflicting migration ids


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

File: src/sentry/models/email.py
Comment (dcramer) [inline src/sentry/models/email.py]: can we add a date_added field here

**Comment by dcramer** [comment]
this is actually good to go at this point right @ehfeng? (i didnt check tests)

**Comment by mattrobenolt** [comment]
I gotta take over and figure out what's causing citext to fail on some of the tests when I get a few spare minutes to poke at it.

File: src/sentry/utils/pytest/sentry.py
Comment (dcramer) [inline src/sentry/utils/pytest/sentry.py]: this code should probably happen after ``fix_south`` is called.

Comment (mattrobenolt) [reply src/sentry/utils/pytest/sentry.py]: It does. Unless you mean immediately after. But either way, I'm gonna see what makes things happy here. This was me just guessing.

Comment (dcramer) [reply src/sentry/utils/pytest/sentry.py]: I mean immediately after, but not sure if it will actually cause issues by keeping it below. Theres sort of a "things should be in an ok state at this point", and I'd just suggest bundling the "database should be ready" parts.

File: src/sentry/models/email.py
Comment (ehfeng) [reply src/sentry/models/email.py]: Added

**Comment by dcramer** [comment]
i'll take this over and figure out why tests are broken real quick

**Comment by dcramer** [comment]
pushed a fix for citext extension

**Comment by ehfeng** [comment]
@dcramer not sure why it's not working, but it looks like things passed: https://travis-ci.org/getsentry/sentry/builds/273059785
Merge?

**Review by dcramer — APPROVED** [review]

**Comment by ehfeng** [comment]
It looks like master is broken right now. I'm going to wait until it's green, rebase, and see what happens. The current failing test seems to have appeared in https://github.com/getsentry/sentry/commit/09883f6ad4331edad68f0fcf172fcc49be96e6e2

**Comment by mattrobenolt** [comment]
You'll need to regenerate migration for new id since it's out of date from current master.

**Comment by ehfeng** [comment]
Regenerated.

**Review by mattrobenolt — APPROVED** [review]

File: src/sentry/models/email.py
Comment (mattrobenolt) [inline src/sentry/models/email.py]: If we don't need an index on this now, we'll probably need to add one later if we try to do anything with this table regarding cleanup.

File: src/sentry/receivers/email.py
Comment (mattrobenolt) [inline src/sentry/receivers/email.py]: I want to note here, that this might be slightly inaccurate due to this not being a case insensitive lookup. You should do `email__iexact`. Because `Email` is case insensitive but `UserEmail` is, you'll get false positives.

File: src/sentry/models/email.py
Comment (ehfeng) [reply src/sentry/models/email.py]: I'm going to punt on this because I'm unsure when we'd need to do such a cleanup.

File: src/sentry/receivers/email.py
Comment (ehfeng) [reply src/sentry/receivers/email.py]: Good point: https://github.com/getsentry/sentry/pull/6070

**Pull request merged**
